### PR TITLE
Rdg/RPC and gateway fixes

### LINKED ISSUE #4988: Connection problems - Windows Server 2008 R2 with gateway
I'm using FreeRDP version 2.0.0-dev4 (76c91ee51) on ThinStation (built on Crux Linux).

On an older build of xfreerdp, I would get frequent disconnects when accessing the Windows server through the gateway with a variety of error codes.  The developers of ThinStation suggested that I pull the latest version from git and compile that into my ThinStation build and now I've got a new issue:

When I try to connect through the gateway with this version, using the following command line:

`xfreerdp /v:(my server) /cert-ignore /f /d:'(my domain)' /g:'(my gateway)' /log-level:TRACE`

Here is the full trace:

```
[14:32:55:312] [5800:5801] [DEBUG][com.freerdp.channels.cliprdr.client] - VirtualChannelEntryEx
[14:32:55:313] [5800:5801] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr
[14:32:55:318] [5800:5801] [DEBUG][com.freerdp.client.x11] - Searching for XInput pointer device
[14:32:55:320] [5800:5801] [DEBUG][com.freerdp.client.x11] - Pointer device: 10
[14:32:55:335] [5800:5801] [DEBUG][com.freerdp.core.nego] - Enabling security layer negotiation: TRUE
[14:32:55:335] [5800:5801] [DEBUG][com.freerdp.core.nego] - Enabling restricted admin mode: FALSE
[14:32:55:335] [5800:5801] [DEBUG][com.freerdp.core.nego] - Enabling RDP security: TRUE
[14:32:55:335] [5800:5801] [DEBUG][com.freerdp.core.nego] - Enabling TLS security: TRUE
[14:32:55:335] [5800:5801] [DEBUG][com.freerdp.core.nego] - Enabling NLA security: TRUE
[14:32:55:335] [5800:5801] [DEBUG][com.freerdp.core.nego] - Enabling NLA extended security: FALSE
[14:32:55:335] [5800:5801] [DEBUG][com.freerdp.core.nego] - state: NEGO_STATE_NLA
[14:32:55:335] [5800:5801] [DEBUG][com.freerdp.core.nego] - Attempting NLA security
[14:32:55:377] [5800:5801] [DEBUG][com.freerdp.core] - connecting to peer 199....16
GatewayUsername: (my username)	
GatewayPassword: 
[14:33:02:889] [5800:5801] [DEBUG][com.winpr.sspi] - InitSecurityInterfaceExA
[14:33:02:889] [5800:5801] [TRACE][com.freerdp.core.gateway.ntlm] - InitializeSecurityContext status SEC_I_CONTINUE_NEEDED [0x00090312]
[14:33:02:931] [5800:5801] [INFO][com.freerdp.core.gateway.rdg] - RD Gateway does not support HTTP transport.
[14:33:02:973] [5800:5801] [DEBUG][com.freerdp.core.gateway.rpc] - VIRTUAL_CONNECTION_STATE_INITIAL
[14:33:02:973] [5800:5801] [DEBUG][com.freerdp.core] - connecting to peer 199.....16
[14:33:02:151] [5800:5801] [DEBUG][com.freerdp.core.gateway.rpc] - CLIENT_IN_CHANNEL_STATE_CONNECTED
[14:33:02:152] [5800:5801] [DEBUG][com.winpr.sspi] - InitSecurityInterfaceExA
[14:33:02:152] [5800:5801] [TRACE][com.freerdp.core.gateway.ntlm] - InitializeSecurityContext status SEC_I_CONTINUE_NEEDED [0x00090312]
[14:33:02:153] [5800:5801] [DEBUG][com.freerdp.core.gateway.rpc] - CLIENT_IN_CHANNEL_STATE_SECURITY
[14:33:02:153] [5800:5801] [DEBUG][com.freerdp.core] - connecting to peer 199.....16
[14:33:02:334] [5800:5801] [DEBUG][com.freerdp.core.gateway.rpc] - CLIENT_OUT_CHANNEL_STATE_CONNECTED
[14:33:02:334] [5800:5801] [DEBUG][com.winpr.sspi] - InitSecurityInterfaceExA
[14:33:02:335] [5800:5801] [TRACE][com.freerdp.core.gateway.ntlm] - InitializeSecurityContext status SEC_I_CONTINUE_NEEDED [0x00090312]
[14:33:02:335] [5800:5801] [DEBUG][com.freerdp.core.gateway.rpc] - CLIENT_OUT_CHANNEL_STATE_SECURITY
[14:33:02:338] [5800:5801] [TRACE][com.freerdp.core.gateway.ntlm] - InitializeSecurityContext status SEC_I_COMPLETE_NEEDED [0x00090313]
[14:33:02:339] [5800:5801] [DEBUG][com.freerdp.core.gateway.rpc] - CLIENT_IN_CHANNEL_STATE_NEGOTIATED
[14:33:02:339] [5800:5801] [DEBUG][com.freerdp.core.gateway.rts] - Sending CONN/B1 RTS PDU
[14:33:02:339] [5800:5801] [DEBUG][com.freerdp.core.gateway.rpc] - CLIENT_IN_CHANNEL_STATE_OPENED
[14:33:02:379] [5800:5801] [TRACE][com.freerdp.core.gateway.ntlm] - InitializeSecurityContext status SEC_I_COMPLETE_NEEDED [0x00090313]
[14:33:02:379] [5800:5801] [DEBUG][com.freerdp.core.gateway.rpc] - CLIENT_OUT_CHANNEL_STATE_NEGOTIATED
[14:33:02:380] [5800:5801] [DEBUG][com.freerdp.core.gateway.rts] - Sending CONN/A1 RTS PDU
[14:33:02:380] [5800:5801] [DEBUG][com.freerdp.core.gateway.rpc] - CLIENT_OUT_CHANNEL_STATE_OPENED
[14:33:02:380] [5800:5801] [DEBUG][com.freerdp.core.gateway.rpc] - VIRTUAL_CONNECTION_STATE_OUT_CHANNEL_WAIT
[14:33:02:425] [5800:5801] [ERROR][com.freerdp.core.gateway.http] - Expected request body too large! (1073741824 bytes) Aborting!
[14:33:02:425] [5800:5801] [ERROR][com.freerdp.core.gateway.tsg] - tsg_check failure
[14:33:02:426] [5800:5801] [ERROR][com.freerdp.core.nego] - Protocol Security Negotiation Failure
[14:33:02:426] [5800:5801] [ERROR][com.freerdp.core] - freerdp_set_last_error ERRCONNECT_SECURITY_NEGO_CONNECT_FAILED [0x0002000C]
[14:33:02:427] [5800:5801] [ERROR][com.freerdp.core.connection] - Error: protocol security negotiation or connection failure
```
Subsequent to that, it fails altogether.  This worked only a day ago on a previous version but disconnected frequently with the same type of error at the end and now it's completely not connecting.

Interestingly enough, I can connect to the server directly without the gateway, for example:

`xfreerdp /v:(my server) /cert-ignore /f /d:'(my domain)' -sec:rdp +nego`

but when I do that, it will connect, work for a bit, and show me all of this:

```
[14:29:19:086] [5596:5597] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr
[14:29:20:781] [5596:5597] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
[14:29:20:781] [5596:5597] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
[14:29:20:838] [5596:5597] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem
[14:29:38:737] [5596:5597] [INFO][com.freerdp.client.x11] - Logon Error Info SESSION_ID [LOGON_MSG_SESSION_CONTINUE]
[14:30:24:588] [5596:5597] [INFO][com.freerdp.core] - ERRINFO_LOGOFF_BY_USER (0x0000000C):The disconnection was initiated by the user logging off their session on the server.
[14:30:24:589] [5596:5597] [ERROR][com.freerdp.core.rdp] - rdp_recv_tpkt_pdu: rdp_recv_deactivate_all() fail
[14:30:24:589] [5596:5597] [ERROR][com.freerdp.core.transport] - transport_check_fds: transport->ReceiveCallback() - -1
Segmentation fault

```
and then it dies with an error when I disconnect.

build config:

```
Build configuration: BUILD_TESTING=OFF BUILTIN_CHANNELS=ON HAVE_AIO_H=1 HAVE_EXECINFO_H=1 HAVE_FCNTL_H=1 HAVE_INTTYPES_H=1 HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK=ON HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL= HAVE_SYSLOG_H=1 HAVE_SYS_EVENTFD_H=1 HAVE_SYS_FILIO_H= HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H= HAVE_SYS_STRTIO_H= HAVE_SYS_TIMERFD_H=1 HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 HAVE_XI_TOUCH_CLASS=1 WITH_ALSA=ON WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CHANNELS_PRINT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUNIT=OFF WITH_CUPS=ON WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DIRECTFB=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_DSP_FFMPEG=OFF WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=ON WITH_FAAD2=OFF WITH_FFMPEG=TRUE WITH_FFMPEG=TRUE WITH_GFX_H264=ON WITH_GPROF=OFF WITH_GSM=OFF WITH_GSSAPI=ON WITH_GSTREAMER_0_10=OFF WITH_GSTREAMER_1_0=ON WITH_ICU=OFF WITH_IPP=OFF WITH_JPEG=ON WITH_LAME=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_NPP=OFF WITH_OPENH264=OFF WITH_OPENSLES=OFF WITH_OPENSSL=ON WITH_OSS=ON WITH_PCSC=ON WITH_PROFILER=OFF WITH_PULSE=ON WITH_SAMPLE=ON WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SERVER=OFF WITH_SERVER_INTERFACE=ON WITH_SMARTCARD_INSPECT=OFF WITH_SSE2=ON WITH_THIRD_PARTY=OFF WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_WAYLAND=OFF WITH_X11=ON WITH_X264=OFF WITH_XCURSOR=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XKBFILE=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XSHM=ON WITH_XV=ON WITH_ZLIB=ON WITH_x264=ON
Build type:          Release
CFLAGS:              -O2 -march=i686 -pipe -march=i686 -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -DWINPR_DLL
Compiler:            GNU, 5.3.0
Target architecture: x86
```

All thoughts welcome!

### PR DESCRIPTION
**Pull request by akallabeth**
Fixes #4988

### CODE REVIEW

File: libfreerdp/core/gateway/rdg.c [hunk 1]
```diff
@@ -306,7 +415,15 @@ static BOOL rdg_set_ntlm_auth_header(rdpNtlm* ntlm, HttpRequest* request)
 	char* base64NtlmToken = NULL;
 
 	if (ntlmToken)
-		base64NtlmToken = crypto_base64_encode(ntlmToken->pvBuffer, ntlmToken->cbBuffer);
+	{
+		if (ntlmToken->cbBuffer > INT_MAX)
+		{
+			free(base64NtlmToken);
```
[the remark this answers is not in the document]
Comment (bmiklautz) [reply libfreerdp/core/gateway/rdg.c]:
```suggestion

```

File: libfreerdp/core/gateway/rdg.c [hunk 2]
```diff
@@ -435,7 +560,7 @@ static BOOL rdg_process_handshake_response(rdpRdg* rdg, wStream* s)
 
 	if (FAILED(errorCode))
 	{
-		WLog_DBG(TAG, "Handshake error %x", errorCode);
+		WLog_DBG(TAG, "Handshake error %"PRIx32, (HRESULT)errorCode);
```
[the remark this answers is not in the document]
Comment (mfleisz) [reply libfreerdp/core/gateway/rdg.c]: Yeah but the cast to HRESULT here seems unnecessary imo ... PRIx32 should be fine with UINT32 as well right?

File: libfreerdp/core/gateway/rdg.c [hunk 3]
```diff
@@ -970,6 +1134,12 @@ static BOOL rdg_process_control_packet(rdpRdg* rdg, int type, size_t packetLengt
 	int status;
 	size_t payloadSize = packetLength - sizeof(RdgPacketHeader);
 
+	if (packetLength < sizeof(RdgPacketHeader))
+		return FALSE;
+
+	if (sizeof(RdgPacketHeader) > INT_MAX)
```
[the remark this answers is not in the document]
Comment (bmiklautz) [reply libfreerdp/core/gateway/rdg.c]: RdgPacketHeader is currently 8 byte.. and has a static size..

File: libfreerdp/core/gateway/rdg.c [hunk 4]
```diff
@@ -1029,10 +1206,13 @@ static int rdg_read_data_packet(rdpRdg* rdg, BYTE* buffer, int size)
 
 	if (!rdg->packetRemainingCount)
 	{
+		if (sizeof(RdgPacketHeader) > INT_MAX)
```
[the remark this answers is not in the document]
Comment (bmiklautz) [reply libfreerdp/core/gateway/rdg.c]: +1

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by bmiklautz** [comment]
I've tested the code quite extensively with all gateway versions I had available.
Basically no problem. However I saw the following error twice but I couldn't reproduce it
```[11:57:30:880] [29153:29154] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpdr
[11:57:30:880] [29153:29154] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpsnd
[11:57:30:880] [29153:29154] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr
[11:57:30:881] [29153:29154] [INFO][com.freerdp.client.x11] - Property 436 does not exist
Certificate details:
  ... snip ...
The above X.509 certificate could not be verified, possibly because you do not have
the CA certificate in your certificate store, or the certificate has expired.
Please look at the OpenSSL documentation on how to add a private CA to the store.
Do you trust the above certificate? (Y/T/N) T
[11:57:33:440] [29153:29154] [ERROR][com.freerdp.core.gateway.rpc] - rpc_client_recv: invalid fragment size: 12832 (max: 4088)
[11:57:33:440] [29153:29154] [ERROR][com.freerdp.core.gateway.rpc] - 0000 48 54 54 50 2f 31 2e 31 20 32 30 30 20 53 75 63 HTTP/1.1 200 Suc
[11:57:33:440] [29153:29154] [ERROR][com.freerdp.core.gateway.tsg] - tsg_check failure
[11:57:33:442] [29153:29154] [ERROR][com.freerdp.core.nego] - Protocol Security Negotiation Failure
[11:57:33:442] [29153:29154] [ERROR][com.freerdp.core] - freerdp_set_last_error ERRCONNECT_SECURITY_NEGO_CONNECT_FAILED [0x0002000C]
```
Command line used:
xfreerdp /g:gw /v:win10host /gu:gu /gp:gp /network:auto /u:user /gt:rpc

**Pull request merged**